code/game/gamemodes/dynamic/dynamic.dm

```diff
@@ -793,6 +812,51 @@ GLOBAL_VAR_INIT(dynamic_forced_threat_level, -1)
 	if (!isnull(threat_log))
 		log_threat(-cost, threat_log, reason)
 
+/// Checks if a high impact ruleset was executed and isn't dead. Only one will run at a time in most circumstances.
+/datum/game_mode/dynamic/proc/high_impact_ruleset_active()
+	. = FALSE
+	for(var/datum/dynamic_ruleset/ruleset in executed_rules)
+		if(!CHECK_BITFIELD(ruleset.flags, HIGH_IMPACT_RULESET))
+			continue
+		if(!high_impact_major_event_occured && ruleset.is_dead())
+			continue
+		return TRUE
+
+/// Checks if the station is considered "intact", for the purpose of rolling rulesets with the INTACT_STATION_RULESET flag.
+/datum/game_mode/dynamic/proc/is_station_intact()
+	if(!COOLDOWN_FINISHED(src, intact_cache_expiry))
+		return cached_station_intact
+	COOLDOWN_START(src, intact_cache_expiry, 5 MINUTES)
+	var/old_cached_station_intact = cached_station_intact
+	cached_station_intact = TRUE
+	var/min_pop = CONFIG_GET(number/dynamic_minimum_living_population)
+	if(min_pop)
+		var/total = 0
+		var/living = 0
+		for(var/mob/living/player in GLOB.player_list)
+			if(!is_station_level(player.z))
+				continue
+			if(player.stat < DEAD)
+				living++
+			total++
+		if(min(total, living) > 0)
+			var/living_percent = living / total
+			if(min_pop > living_percent)
+				cached_station_intact = FALSE
+				if(old_cached_station_intact)
+					log_game("DYNAMIC: Station has a living population of [PERCENT(living_percent)]%, which is lower than the minimum living population of [PERCENT(min_pop)]%")
+	var/min_integrity = CONFIG_GET(number/dynamic_minimum_station_integrity)
+	if(min_integrity)
+		var/datum/station_state/current_state = new /datum/station_state()
```

Comment:
is there not already one of these stored somewhere for say, the stat panel or roundend panel? Not sure how it works just checking